app/services/tagging_history/link_changes.rb

```diff
@@ -0,0 +1,45 @@
+module TaggingHistory
+  class LinkChanges
+    def initialize(params)
+      @params = params
+    end
+
+    def changes
+      @_changes ||= link_changes_from_publishing_api.map do |change|
+        user = User.find_by_uid(change[:user_uid])
+
+        if user.present?
+          change[:user_uid] = user.uid
+          change[:user_name] = user.name
+          change[:organisation] = user.organisation_slug.try do |slug|
+            slug.capitalize.tr('-', ' ')
+          end
+        end
+
+        change
+      end
+    end
+
+    def filter_by_user_options
+      options = {}
+
+      changes.each do |change|
```

Comment:
could filter the changes instead and then iterate.

`changes.keep_if {|change| change.key? :user_id }.each...`

Could also use inject/each_with_object to build the options:

```
filtered_changes.each_with_object({}) do |change, options|
  options[change[:user_name]] = change[:user_uid]
end
```

Reply:
I don't particularly like the chaining with blocks, mostly because it looks really messy, I do really like the use of each_with_object (or reduce as I would normally call it...), so I've switched to that here 3f7304a